File: dotnet/src/SatisfactoryExtractor/GameFinderOptions.cs
namespace Rosser.SatisfactoryExtractor;

using Rosser.SatisfactoryExtractor.Validation;

public class GameFinderOptions
{
    /// <summary>Satisfactory update channel</summary>
    public Channel Channel { get; set; } = Channel.Experimental;

    /// <summary>true to only look for the game in Steam paths</summary>
    public bool OnlySteam { get; set; }

    /// <summary>true to only look for the game in Epic Games paths</summary>
    public bool OnlyEpic { get; set; }

    /// <summary>
    /// Gets a value indicating whether the options are valid.
    /// </summary>
    public bool IsValid => new FluentValidator()
        .Single(new[] {
            new FluentValidatorTest { Name = nameof(this.OnlyEpic), Test = () => this.OnlyEpic },
            new FluentValidatorTest { Name = nameof(this.OnlySteam), Test = () => this.OnlySteam },
        }).IsValid;
}

File: typescript/src/game-finder-options.ts
import { FluentValidator } from './fluent-validator';

export type Channel = 'experimental' | 'earlyaccess';

export const isChannel = (obj: unknown): obj is Channel => {
  return obj === 'experimental' || obj === 'earlyaccess';
};

export interface GameFinderOptions {
  /** Satisfactory update channel */
  channel: Channel;

  /** true to only look for the game in Steam paths */
  onlySteam?: boolean;

  /** true to only look for the game in Epic Games paths */
  onlyEpic?: boolean;
}

export const isGameFinderOptions = (obj: unknown): obj is GameFinderOptions => {
  const options = obj as GameFinderOptions;
  return validate(options).errors.length == 0;
};

const validate = (options: GameFinderOptions) => {
  const validator = new FluentValidator()
    .mutuallyExclusive([
      { name: 'onlyEpic', test: () => !!options.onlyEpic && options.onlyEpic },
      { name: 'onlySteam', test: () => !!options.onlySteam && options.onlySteam }
    ])
    .oneOf('channel', options.channel, ['experimental', 'earlyaccess']);

  if (process.platform === 'linux') {
    validator.fail('Not supported on linux');
  }

  return validator;
};